File: app/src/main/java/fhu/bughousechess/Piece.java
package fhu.bughousechess;

import android.widget.ImageView;

import java.util.Set;

public abstract class Piece
{
    public String color;
    public String type;
    public boolean wasPawn;
    public String backgroundColor;
    public boolean empty = true;

    public abstract int getResID();
    public abstract Set<Move> getMoves(ImageView[][] board, Piece[][] positions, int x, int y);
    public boolean isOpposite(Piece piece)
    {
        if (color.equals("white"))
        return piece.color.equals("black");
        if (color.equals("black"))
            return piece.color.equals("white");
        return false;

    }


}

File: app/src/main/java/fhu/bughousechess/pieces/Piece.kt
package fhu.bughousechess.pieces

import fhu.bughousechess.Move
import java.util.HashSet

abstract class Piece {
    @JvmField
    var color: String? = null
    @JvmField
    var type: String? = null
    @JvmField
    var wasPawn = false
    @JvmField
    var empty = true
    abstract fun getMoves(
        positions: Array<Array<Piece>>,
        x: Int,
        y: Int,
        boardNumber: Int
    ): Set<Move?>?

    fun isOpposite(piece: Piece): Boolean {
        if (color == "white") return piece.color == "black"
        return if (color == "black") piece.color == "white" else false
    }

    open fun getRosterMoves(positions: Array<Array<Piece>>, rosterp: Array<Piece>, i: Int): Set<Move> {
        val moves: MutableSet<Move> = HashSet()
        for (x in 0..7) {
            for (y in 0..7) {
                if (positions[x][y].empty) {
                    moves.add(Move(positions, rosterp, i, x, y, "roster"))
                }
            }
        }
        return moves
    }
}